core/src/test/scala/org/apache/spark/sql/delta/DeltaSuite.scala

```diff
@@ -596,6 +596,144 @@ class DeltaSuite extends QueryTest
+        .write
+        .format("delta")
+        .partitionBy("part")
+        .mode("append")
+        .save(tempDir.getCanonicalPath)
+
+      Seq(1, 5).toDF
+        .withColumn("part", $"value" % 2)
+        .write
+        .format("delta")
+        .partitionBy("part")
+        .mode("overwrite")
+        .option("partitionOverwriteMode", "dynamic")
+        .save(tempDir.getCanonicalPath)
+      checkDatasetUnorderly(data.select("value").as[Int], 1, 2, 5)
+    }
+
+    withTempDir { tempDir =>
+      def data: DataFrame = spark.read.format("delta").load(tempDir.toString)
+
+      Seq(("a", "x"), ("b", "y"), ("c", "x")).toDF("value", "part")
+        .write
+        .format("delta")
+        .partitionBy("part")
+        .mode("append")
+        .save(tempDir.getCanonicalPath)
+
+      Seq(("a", "x"), ("d", "x")).toDF("value", "part")
+        .write
+        .format("delta")
+        .partitionBy("part")
+        .mode("overwrite")
+        .option("partitionOverwriteMode", "dynamic")
+        .save(tempDir.getCanonicalPath)
+      checkDatasetUnorderly(data.select("value").as[String], "a", "b", "d")
+    }
+
+    withTempDir { tempDir =>
+      // overwrites nothing
+
+      def data: DataFrame = spark.read.format("delta").load(tempDir.toString)
+
+      Seq(("a", "x"), ("b", "y"), ("c", "x")).toDF("value", "part")
+        .write
+        .format("delta")
+        .partitionBy("part")
+        .mode("append")
+        .save(tempDir.getCanonicalPath)
+
+      Seq(("d", "z")).toDF("value", "part")
+        .write
+        .format("delta")
+        .partitionBy("part")
+        .mode("overwrite")
+        .option("partitionOverwriteMode", "dynamic")
+        .save(tempDir.getCanonicalPath)
+      checkDatasetUnorderly(data.select("value", "part").as[(String, String)],
+        ("a", "x"), ("b", "y"), ("c", "x"), ("d", "z"))
+    }
+
+    withTempDir { tempDir =>
+      // multiple partition columns
+
+      def data: DataFrame = spark.read.format("delta").load(tempDir.toString)
+
+      Seq(("a", "x", 1), ("b", "y", 2), ("c", "x", 3)).toDF("part1", "part2", "value")
+        .write
+        .format("delta")
+        .partitionBy("part1", "part2")
+        .mode("append")
+        .save(tempDir.getCanonicalPath)
+
+      Seq(("a", "x", 4), ("d", "x", 5)).toDF("part1", "part2", "value")
+        .write
+        .format("delta")
+        .partitionBy("part1", "part2")
+        .mode("overwrite")
+        .option("partitionOverwriteMode", "dynamic")
+        .save(tempDir.getCanonicalPath)
+      checkDatasetUnorderly(data.select("part1", "part2", "value").as[(String, String, Int)],
+        ("a", "x", 4), ("b", "y", 2), ("c", "x", 3), ("d", "x", 5))
+    }
+  }
+
+  test("batch write: append, dynamic partition overwrite with replaceWhere") {
+    withTempDir { tempDir =>
+      def data: DataFrame = spark.read.format("delta").load(tempDir.toString)
+
+      Seq((1, "x"), (2, "y"), (3, "z")).toDF("value", "sub")
```

Comment:
whats the purpose of this added column `sub`

Reply:
sub is added so we can test partitioning on 2 columns

Reply:
this detail is not clear. depending on what you think about this [question](https://github.com/delta-io/delta/pull/371/files#r886231061, we have to either modify / add tests related to replaceWhere and in that case, it would be good to have an inline comment explaining this. also in that case, `sub` is not a great name, might as well do `part1` and `part2`

Reply:
now checking that error is throw here. also renamed columns to `part1` and `part2`.